test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,861 @@
+    auto logger = util::Logger::get_default_logger();
+    static std::unique_ptr<FLXSyncTestHarness> harness;
+
+    // 150 emps, 25 mgrs, 10 dirs
+    // 10 objects before flush
+    // 1536 download soft max bytes
+    TestParams params{};
+    params.max_download_bytes = 1536;
+    if (!harness) {
+        harness = setup_harness("flx_role_change_during_bs", params);
+    }
+    REQUIRE(harness);
+
+    // Get the current rules so it can be updated during the test
+    auto& app_session = harness->session().app_session();
+    auto default_rule = app_session.admin_api.get_default_rule(app_session.server_app_id);
+
+    // Make sure the rules are reset back to the original value (all records allowed)
+    update_role(default_rule, true);
+    logger->debug("ROLE CHANGE: Initial rule definitions: %1", default_rule);
+    REQUIRE(app_session.admin_api.update_default_rule(app_session.server_app_id, default_rule));
+
+    enum BootstrapTestState {
+        not_ready,
+        start,
+        ident_sent,
+        reconnect_received,
+        downloading,
+        downloaded,
+        integrating,
+        integration_complete,
+        complete
+    };
+
+    BootstrapTestState update_role_state = BootstrapTestState::not_ready;
+    int update_msg_count = -1;
+    int bootstrap_count = 0;
+    int bootstrap_msg_count = 0;
+    bool session_restarted = false;
+    std::optional<util::Future<void>> role_change_future;
+    TestingStateMachine<BootstrapTestState> bootstrap_state(BootstrapTestState::not_ready);
+
+    auto setup_config_callbacks = [&](SyncTestFile& config) {
+        // Use the sync client event hook to check for the error received and for tracking
+        // download messages and bootstraps
+        config.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession>,
+                                                            const SyncClientHookData& data) {
+            bootstrap_state.transition_with([&](BootstrapTestState cur_state) -> std::optional<BootstrapTestState> {
+                using BatchState = sync::DownloadBatchState;
+                using Event = SyncClientHookEvent;
+                // Keep track of the number of bootstraps that have occurred, regardless of cur state
+                if (data.event == Event::BootstrapProcessed) {
+                    bootstrap_count++;
+                }
+
+                // Has the test started?
+                if (cur_state == BootstrapTestState::not_ready)
+                    return std::nullopt;
+
+                std::optional<BootstrapTestState> new_state = std::nullopt;
+
+                switch (data.event) {
+                    case Event::IdentMessageSent:
+                        new_state = BootstrapTestState::ident_sent;
+                        break;
+
+                    case Event::ErrorMessageReceived:
+                        REQUIRE(data.error_info);
+                        if (data.error_info->raw_error_code == 200) {
```

Comment:
are there other errors we could receive here that we don't want to track?

Reply:
Not really - changed the `if` to a `REQUIRE()`